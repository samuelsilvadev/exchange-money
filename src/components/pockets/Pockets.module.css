.pockets {
	display: flex;
	flex-wrap: wrap;
	justify-content: space-between;
	align-items: flex-start;
	align-content: flex-start;
}

.title {
	margin-bottom: 5rem;
	width: 100%;
}

.pocketSelect {
	width: 10rem;
}

.pocketBallance {
	font-size: 1.6rem;
	text-align: right;
	padding: 0.5rem 0 0.25rem;
}

.pocketBallanceDescription {
	font-size: 1.4rem;
	color: var(--gray);
}
