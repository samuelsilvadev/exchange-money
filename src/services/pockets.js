export const initialState = [
	{
		label: 'USD',
		value: 1000,
	},
	{
		label: 'GBP',
		value: 5000,
	},
	{
		label: 'EUR',
		value: 2000,
	},
];

export function getPockets() {
	return Promise.resolve(initialState);
}
