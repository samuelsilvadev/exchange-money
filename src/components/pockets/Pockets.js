import React, { useEffect, useState } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

import { usePocketsContext } from '../../providers/pockets/PocketsProvider';
import Select from '../select/Select';

import styles from './Pockets.module.css';

function Pockets(props) {
	const { className } = props;

	const { result: pockets } = usePocketsContext();

	const [selectedPocket, setSelectedPocket] = useState(pockets?.[0]);

	useEffect(() => {
		if (pockets.length > 0) {
			const updatedPocket = pockets.find(
				(pocket) => pocket.label === selectedPocket?.label
			);

			setSelectedPocket(updatedPocket ?? pockets[0]);
		}
	}, [pockets, selectedPocket]);

	const handleChange = (event) => {
		const label = event.target.value;

		const nextSelectedPocket = pockets.find(
			(pocket) => pocket.label === label
		);

		setSelectedPocket(nextSelectedPocket);
	};

	return (
		<section
			className={classnames(styles.pockets, { [className]: className })}
		>
			<h1 className={styles.title}>Pockets</h1>
			<Select
				id="pocket"
				label="Pocket"
				className={styles.pocketSelect}
				onBlur={handleChange}
				onChange={handleChange}
			>
				{pockets.map(({ label }) => {
					return (
						<option value={label} key={label}>
							{label}
						</option>
					);
				})}
			</Select>
			<div>
				<p className={styles.pocketBallance}>
					{selectedPocket?.value ?? '-'}
				</p>
				<p className={styles.pocketBallanceDescription}>
					Total Ballance
				</p>
			</div>
		</section>
	);
}

Pockets.propTypes = {
	className: PropTypes.string,
};

export default Pockets;
