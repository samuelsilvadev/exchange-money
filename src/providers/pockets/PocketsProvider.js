import React, {
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { getPockets } from '../../services/pockets';

/**
 * A pocket description.
 *
 * @typedef {Object} Pocket
 * @property {string} label
 * @property {number} value
 */

/**
 * The pocket context description.
 *
 * @typedef {Object} PocketsContext
 * @property {boolean} isLoading
 * @property {boolean} hasError
 * @property {Array<Pocket>} result
 * @property {Function} transfer
 */

const checkForErrorOnTransferValues = ({ pocket, value }) => {
	if (pocket.value === 0) {
		return new TypeError(`Your ${pocket.label} pocket is empty`);
	}

	if (pocket.value < value) {
		return new TypeError(
			`Insufficient money on your ${pocket.label} pocket`
		);
	}

	return null;
};

const PocketsContext = createContext();

export function PocketsProvider(props) {
	const [pockets, setPockets] = useState([]);
	const [isLoading, setIsLoading] = useState(false);
	const [hasError, setHasError] = useState(false);

	useEffect(() => {
		setIsLoading(true);
		setHasError(false);

		getPockets()
			.then((data) => {
				if (data) {
					setPockets(data);
				}
			})
			.catch(() => setHasError(true))
			.finally(() => setIsLoading(false));
	}, []);

	const transfer = useCallback(
		(from, to) => {
			setIsLoading(true);
			setHasError(false);

			return new Promise((resolve, reject) => {
				const updatedPockets = pockets.reduce(
					(updatedPockets, currentPocket) => {
						const pocket = { ...currentPocket };

						if (pocket.label === from.currency) {
							const parsedFromMoneyAmount = parseFloat(
								from.moneyAmount
							);
							const error = checkForErrorOnTransferValues({
								pocket,
								value: parsedFromMoneyAmount,
							});

							if (error) {
								return reject(error);
							}

							pocket.value -= parsedFromMoneyAmount;
						} else if (pocket.label === to.currency) {
							pocket.value += parseFloat(to.moneyAmount);
						}

						updatedPockets.push(pocket);

						return updatedPockets;
					},
					[]
				);

				setPockets(updatedPockets);

				resolve();
			})
				.catch((error) => {
					setHasError(true);

					return error;
				})
				.finally(() => setIsLoading(false));
		},
		[pockets]
	);

	const context = useMemo(() => {
		return {
			isLoading,
			hasError,
			result: pockets,
			transfer,
		};
	}, [isLoading, hasError, pockets, transfer]);

	return <PocketsContext.Provider value={context} {...props} />;
}

/**
 * Get pocket values from global storage.
 *
 * @returns {PocketsContext} - Result set of pockets objects.
 */
export function usePocketsContext() {
	const pockets = useContext(PocketsContext);

	if (typeof pockets === 'undefined') {
		throw new TypeError(
			'To use the PocketsContext correctly you should wrap your component with <PocketsProvider />'
		);
	}

	return pockets;
}
