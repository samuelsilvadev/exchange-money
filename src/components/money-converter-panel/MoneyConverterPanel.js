import React from 'react';
import PropTypes from 'prop-types';

import { usePocketsContext } from '../../providers/pockets/PocketsProvider';
import Select from '../select/Select';

import styles from './MoneyConverterPanel.module.css';

function MoneyConverterPanel(props) {
	const {
		label,
		className,
		handlers: {
			currency: currencyHandler,
			moneyAmount: moneyAmountHandler,
		} = {},
		values: { currency: currencyValue, moneyAmount: moneyAmountValue } = {},
	} = props;

	const pockets = usePocketsContext();

	return (
		<div className={className}>
			<Select
				className={styles.select}
				label="Currency"
				id={`converter-panel-${label}`}
				onChange={currencyHandler}
				onBlur={currencyHandler}
				value={currencyValue}
			>
				{pockets.map(({ label }) => {
					return (
						<option value={label} key={label}>
							{label}
						</option>
					);
				})}
			</Select>
			<div className={styles.inputWrapper}>
				<label className={styles.label}>{label}</label>
				<input
					className={styles.input}
					onChange={moneyAmountHandler}
					value={moneyAmountValue}
					type="text"
				/>
			</div>
		</div>
	);
}

MoneyConverterPanel.propTypes = {
	className: PropTypes.string,
	label: PropTypes.string.isRequired,
	values: PropTypes.shape({
		currency: PropTypes.string,
		moneyAmount: PropTypes.string,
	}),
	handlers: PropTypes.shape({
		currency: PropTypes.func,
		moneyAmount: PropTypes.func,
	}),
};

export default MoneyConverterPanel;
