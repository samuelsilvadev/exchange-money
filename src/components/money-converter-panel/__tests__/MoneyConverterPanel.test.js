import React from 'react';
import { render } from '@testing-library/react';

import { PocketsProvider } from '../../../providers/pockets/PocketsProvider';
import MoneyConverterPanel from '../MoneyConverterPanel';

describe('<MoneyConverterPanel />', () => {
	it('should render correctly', () => {
		const { container, getByLabelText, getByText } = render(
			<PocketsProvider>
				<MoneyConverterPanel label="You get" />
			</PocketsProvider>
		);

		expect(getByLabelText('Currency')).toBeVisible();
		expect(getByText('You get')).toBeVisible();

		expect(container.firstChild).toMatchSnapshot();
	});
});
