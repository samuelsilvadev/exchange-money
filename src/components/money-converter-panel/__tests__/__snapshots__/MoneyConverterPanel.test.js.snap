// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MoneyConverterPanel /> should render correctly 1`] = `
<div>
  <div
    class="selectWrapper select"
  >
    <label
      class="selectLabel"
      for="converter-panel-You get"
    >
      Currency
    </label>
    <div
      class="selectIconWrapper"
    >
      <select
        class="select"
        id="converter-panel-You get"
      >
        <option
          value="USD"
        >
          USD
        </option>
        <option
          value="GBP"
        >
          GBP
        </option>
        <option
          value="EUR"
        >
          EUR
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="selectIcon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 15l-4.243-4.243 1.415-1.414L12 12.172l2.828-2.829 1.415 1.414z"
        />
      </svg>
    </div>
  </div>
  <div
    class="inputWrapper"
  >
    <label
      class="label"
    >
      You get
    </label>
    <input
      class="input"
      type="text"
      value=""
    />
  </div>
</div>
`;
